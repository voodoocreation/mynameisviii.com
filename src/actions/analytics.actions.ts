import actionCreatorFactory from "typescript-fsa";

const createAction = actionCreatorFactory("ANALYTICS");

export const trackEvent = createAction<{
  event: string;
  [index: string]: any;
}>("TRACK_EVENT");
