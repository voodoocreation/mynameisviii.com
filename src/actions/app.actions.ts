import actionCreatorFactory from "typescript-fsa";

const createAction = actionCreatorFactory("APP");

export const initApp = createAction.async<{ locale?: string }, {}>("INIT");

export const toggleNavigation = createAction("TOGGLE_NAVIGATION");

export const setOnlineStatus = createAction<boolean>("SET_ONLINE_STATUS");

export const setHasNewVersion = createAction<boolean>("SET_HAS_NEW_VERSION");

export const setCurrentRoute = createAction<string>("SET_CURRENT_ROUTE");

export const changeRoute = createAction.async<string, {}>("CHANGE_ROUTE");
