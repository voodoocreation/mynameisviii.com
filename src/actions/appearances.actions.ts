import actionCreatorFactory from "typescript-fsa";

import { IAppearance, IDynamoResponse, ILatLng } from "../models/root.models";

const createAction = actionCreatorFactory("APPEARANCES");

export const fetchAppearances = createAction.async<
  {},
  IDynamoResponse<IAppearance, "startingAt">
>("FETCH");

export const fetchMoreAppearances = createAction.async<
  {},
  IDynamoResponse<IAppearance, "startingAt">
>("FETCH_MORE");

export const fetchAppearanceBySlug = createAction.async<string, IAppearance>(
  "FETCH_BY_SLUG"
);

export const setCurrentAppearanceSlug = createAction<string>(
  "SET_CURRENT_SLUG"
);

export const geocodeCurrentAppearanceAddress = createAction.async<{}, ILatLng>(
  "GEOCODE_CURRENT_ADDRESS"
);
