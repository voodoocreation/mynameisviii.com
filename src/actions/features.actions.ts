import actionCreatorFactory from "typescript-fsa";

const actionCreator = actionCreatorFactory("FEATURES");

export const addFeatures = actionCreator<PLAddFeatures>("ADD");

export const removeFeatures = actionCreator<PLRemoveFeatures>("REMOVE");
