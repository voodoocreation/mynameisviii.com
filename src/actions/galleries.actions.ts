import actionCreatorFactory from "typescript-fsa";

const actionCreator = actionCreatorFactory("GALLERIES");

export const fetchGalleries = actionCreator.async<
  {},
  PLFetchGalleriesDone,
  PLFetchGalleriesFailed
>("FETCH");

export const fetchMoreGalleries = actionCreator.async<
  {},
  PLFetchGalleriesDone,
  PLFetchGalleriesFailed
>("FETCH_MORE");

export const fetchGalleryBySlug = actionCreator.async<
  PLFetchGalleryBySlugStarted,
  PLFetchGalleryBySlugDone,
  PLFetchGalleryBySlugFailed
>("FETCH_BY_SLUG");

export const setCurrentGallerySlug = actionCreator<PLSetCurrentGallerySlug>(
  "SET_CURRENT_SLUG"
);
