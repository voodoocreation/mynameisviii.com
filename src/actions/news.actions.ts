import actionCreatorFactory from "typescript-fsa";

const actionCreator = actionCreatorFactory("NEWS");

export const fetchLatestNews = actionCreator.async<
  {},
  PLFetchLatestNewsDone,
  PLFetchLatestNewsFailed
>("FETCH");

export const fetchMoreLatestNews = actionCreator.async<
  {},
  PLFetchLatestNewsDone,
  PLFetchLatestNewsFailed
>("FETCH_MORE");

export const fetchNewsArticleBySlug = actionCreator.async<
  PLFetchNewsArticleBySlugStarted,
  PLFetchNewsArticleBySlugDone,
  PLFetchNewsArticleBySlugFailed
>("FETCH_BY_SLUG");

export const setCurrentNewsArticleSlug = actionCreator<
  PLSetCurrentNewsArticleSlug
>("SET_CURRENT_SLUG");
