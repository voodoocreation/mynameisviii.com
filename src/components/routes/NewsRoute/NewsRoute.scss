@import "../../../scss/variables";

.NewsRoute--listings {
  padding-bottom: $spacing;
}
