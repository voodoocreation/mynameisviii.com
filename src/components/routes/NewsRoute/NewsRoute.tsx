import cn from "classnames";
import Head from "next/head";
import * as React from "react";
import { FormattedMessage, InjectedIntl } from "react-intl";
import { connect } from "react-redux";
import { bindActionCreators, Dispatch } from "redux";
import { ActionCreator } from "typescript-fsa";

import injectIntlIntoPage from "../../../helpers/injectIntlIntoPage";
import { absUrl } from "../../../transformers/transformData";
import ButtonBar from "../../presentation/ButtonBar/ButtonBar";
import LoadButton from "../../presentation/LoadButton/LoadButton";
import NewsListing from "../../presentation/NewsListing/NewsListing";
import NoResults from "../../presentation/NoResults/NoResults";
import PageHeader from "../../presentation/PageHeader/PageHeader";

import * as actions from "../../../actions/root.actions";
import * as selectors from "../../../selectors/root.selectors";

interface IStoreProps {
  articles: INewsArticle[];
  articlesCount: number;
  hasAllNewsArticles: boolean;
  isLoading: boolean;
}

interface IDispatchProps {
  fetchLatestNews: ActionCreator<{}>;
  fetchMoreLatestNews: ActionCreator<{}>;
}

interface IProps extends IStoreProps, IDispatchProps {
  intl: InjectedIntl;
}

interface IState {
  loadedListings: {
    [index: string]: boolean;
  };
}

class NewsRoute extends React.Component<IProps, IState> {
  public static async getInitialProps(props: any) {
    const { isServer, store } = props.ctx;

    const state = store.getState();

    if (isServer || selectors.getNewsArticlesCount(state) === 0) {
      store.dispatch(actions.fetchLatestNews.started({}));
    } else if (!selectors.getHasAllNewsArticles(state)) {
      store.dispatch(actions.fetchMoreLatestNews.started({}));
    }
  }

  public readonly state = {
    loadedListings: {}
  };

  public render() {
    const { articles, articlesCount, hasAllNewsArticles } = this.props;
    const { formatMessage } = this.props.intl;

    const hasLoadedAllListings =
      articlesCount === Object.keys(this.state.loadedListings).length;

    const isLoading = this.props.isLoading || !hasLoadedAllListings;

    const loadMoreButton = hasAllNewsArticles ? null : (
      <LoadButton isLoading={isLoading} onLoad={this.onLoadMore}>
        <FormattedMessage id="LOAD_MORE" />
      </LoadButton>
    );

    const pageTitle = formatMessage({ id: "NEWS_TITLE" });
    const pageDescription = formatMessage({ id: "NEWS_DESCRIPTION" });

    return (
      <article className={cn("NewsRoute", { hasLoadedAllListings })}>
        <Head>
          <title>
            {pageTitle}
            {" Â· "}
            {formatMessage({ id: "BRAND_NAME" })}
          </title>

          <meta content={pageDescription} name="description" />

          <meta property="og:title" content={pageTitle} />
          <meta property="og:description" content={pageDescription} />
          <meta
            property="og:image"
            content="https://s3.amazonaws.com/mynameisviii-static/news-og.jpg"
          />
          <meta property="og:url" content={absUrl("/news")} />
          <meta property="og:type" content="website" />
        </Head>

        <PageHeader>
          <FormattedMessage id="NEWS_TITLE" />
        </PageHeader>

        {articlesCount > 0 ? (
          <section className="NewsListings">
            {articles.map(article => (
              <NewsListing
                {...article}
                key={article.slug}
                onLoad={this.onListingLoad}
              />
            ))}
          </section>
        ) : null}

        {hasAllNewsArticles && articlesCount === 0 ? (
          <NoResults>
            <p>
              <FormattedMessage id="NO_NEWS" />
            </p>
          </NoResults>
        ) : null}

        <ButtonBar>{loadMoreButton}</ButtonBar>
      </article>
    );
  }

  private onListingLoad = (slug: string) => {
    this.setState({
      loadedListings: {
        ...this.state.loadedListings,
        [slug]: true
      }
    });
  };

  private onLoadMore = () => {
    this.props.fetchMoreLatestNews({});
  };
}

const mapStateToProps = (state: any) => ({
  articles: selectors.getNewsArticlesAsArray(state),
  articlesCount: selectors.getNewsArticlesCount(state),
  hasAllNewsArticles: selectors.getHasAllNewsArticles(state),
  isLoading: selectors.getNewsIsLoading(state)
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      fetchLatestNews: actions.fetchLatestNews.started,
      fetchMoreLatestNews: actions.fetchMoreLatestNews.started
    },
    dispatch
  );

export default injectIntlIntoPage(
  connect<IStoreProps, IDispatchProps>(
    mapStateToProps,
    mapDispatchToProps
  )(NewsRoute)
);
