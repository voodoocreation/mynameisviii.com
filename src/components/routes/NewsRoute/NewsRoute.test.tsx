import * as actions from "../../../actions/root.actions";
import { newsArticle } from "../../../models/root.models";
import MockPageContext from "../../../utilities/MockPageContext";
import WrapperWithRedux from "../../../utilities/WrapperWithRedux";
import NewsRoute from "./NewsRoute";

const item1 = newsArticle({ slug: "test-1" });
const item2 = newsArticle({ slug: "test-2" });

const component = new WrapperWithRedux(NewsRoute);

describe("[routes] <NewsRoute />", () => {
  describe("getInitialProps", () => {
    const context = new MockPageContext();

    it("dispatches actions.fetchLatestNews.started when rendering on the server", async () => {
      await NewsRoute.getInitialProps(context.toObject(true));

      expect(
        context.reduxHistory.filter(actions.fetchLatestNews.started.match)
      ).toHaveLength(1);
    });

    it("dispatches actions.fetchLatestNews.started when the store is empty", async () => {
      await NewsRoute.getInitialProps(context.toObject());

      expect(
        context.reduxHistory.filter(actions.fetchLatestNews.started.match)
      ).toHaveLength(1);
    });

    it("dispatches actions.fetchMoreLatestNews.started when there are more items to fetch", async () => {
      await NewsRoute.getInitialProps(
        context
          .withReduxState({
            news: {
              hasAllItems: false,
              items: {
                [item1.slug]: item1
              }
            }
          })
          .toObject()
      );

      expect(
        context.reduxHistory.filter(actions.fetchMoreLatestNews.started.match)
      ).toHaveLength(1);
    });

    it("doesn't fetch anything when all items are in the store", async () => {
      await NewsRoute.getInitialProps(
        context
          .withReduxState({
            news: {
              hasAllItems: true,
              items: {
                [item1.slug]: item1
              }
            }
          })
          .toObject()
      );

      expect(
        context.reduxHistory.filter(actions.fetchLatestNews.started.match)
      ).toHaveLength(0);
      expect(
        context.reduxHistory.filter(actions.fetchMoreLatestNews.started.match)
      ).toHaveLength(0);
    });
  });

  describe("when all data is in the store", () => {
    const wrapper = component
      .withReduxState({
        news: {
          hasAllItems: true,
          items: {
            [item1.slug]: item1,
            [item2.slug]: item2
          }
        }
      })
      .mount();

    it("renders the listings section", () => {
      expect(wrapper.find(".NewsRoute--listings")).toHaveLength(1);
    });

    it("renders all listings", () => {
      expect(wrapper.find("NewsListing")).toHaveLength(2);
    });

    it("matches snapshot", () => {
      expect(wrapper.render()).toMatchSnapshot();
    });
  });

  describe("when no data is in the store", () => {
    const wrapper = component
      .withReduxState({
        news: {
          hasAllItems: true,
          items: {}
        }
      })
      .mount();

    it("doesn't render the listings section", () => {
      expect(wrapper.find(".NewsRoute--listings")).toHaveLength(0);
    });

    it("renders the no results message", () => {
      expect(wrapper.find("NoResults")).toHaveLength(1);
    });
  });

  it("renders the offline notice when there's an error", () => {
    const wrapper = component
      .withReduxState({
        news: {
          hasError: true
        }
      })
      .mount();

    expect(wrapper.find("OfflineNotice")).toHaveLength(1);
  });

  it("renders with the 'load more' button when hasAllItems is false", () => {
    const wrapper = component
      .withReduxState({
        news: {
          hasAllItems: false,
          items: {
            [item1.slug]: item1
          }
        }
      })
      .mount();

    expect(wrapper.find("LoadButton")).toHaveLength(1);
  });

  it("renders without the 'load more' button when hasAllItems is true", () => {
    const wrapper = component
      .withReduxState({
        news: {
          hasAllItems: true
        }
      })
      .mount();

    expect(wrapper.find("LoadButton")).toHaveLength(0);
  });

  it("loads more listings when the 'load more' button is clicked", () => {
    const wrapper = component.mount();

    wrapper.find("LoadButton").simulate("click");

    expect(
      component.reduxHistory.filter(actions.fetchMoreLatestNews.started.match)
    );
  });

  it("updates `loadedListings` state after a listing has loaded", () => {
    const wrapper = component
      .withReduxState({
        news: {
          items: {
            [item1.slug]: item1
          }
        }
      })
      .mount();

    // @ts-ignore-next-line
    wrapper.find("NewsListing").prop("onLoad")();

    expect(wrapper.update().find(".hasLoadedAllListings")).toHaveLength(1);
  });
});
