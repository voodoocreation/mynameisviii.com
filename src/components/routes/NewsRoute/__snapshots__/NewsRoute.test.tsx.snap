// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[routes] <NewsRoute /> when all data is in the store matches snapshot 1`] = `
<article
  class="NewsRoute"
>
  <header
    class="PageHeader"
  >
    <h1>
      News
    </h1>
  </header>
  <section
    class="NewsRoute--listings"
  >
    <article
      class="NewsListing"
    >
      <a
        href="/news/test-1"
      >
        <header
          class="NewsListing--header"
        >
          <h2>
            <div
              class="NewsArticleType NewsArticleType-news"
              title="News"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M23 12l-2.44-2.78.34-3.68-3.61-.82-1.89-3.18L12 3 8.6 1.54 6.71 4.72l-3.61.81.34 3.68L1 12l2.44 2.78-.34 3.69 3.61.82 1.89 3.18L12 21l3.4 1.46 1.89-3.18 3.61-.82-.34-3.68L23 12zm-10 5h-2v-2h2v2zm0-4h-2V7h2v6z"
                />
              </svg>
            </div>
            <span />
          </h2>
        </header>
        <p>
          <span />
        </p>
      </a>
      <script
        type="application/ld+json"
      >
        {"@id":"http://localhost:/news/test-1","@type":"NewsArticle","articleBody":"","author":{"@type":"Person","name":""},"dateModified":"2017-12-31T11:00:00.000Z","datePublished":"2017-12-31T11:00:00.000Z","description":"","headline":"","image":"","mainEntityOfPage":{"@id":"http://localhost:/news/test-1"},"name":"","publisher":{"@type":"LocalBusiness","email":"mgmt@mynameisviii.com","logo":{"@type":"ImageObject","url":"https://s3.amazonaws.com/mynameisviii-static/voodoo-creation-records-logo.png"}},"text":"","thumbnailUrl":"","@context":"http://schema.org/"}
      </script>
    </article>
    <article
      class="NewsListing"
    >
      <a
        href="/news/test-2"
      >
        <header
          class="NewsListing--header"
        >
          <h2>
            <div
              class="NewsArticleType NewsArticleType-news"
              title="News"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M23 12l-2.44-2.78.34-3.68-3.61-.82-1.89-3.18L12 3 8.6 1.54 6.71 4.72l-3.61.81.34 3.68L1 12l2.44 2.78-.34 3.69 3.61.82 1.89 3.18L12 21l3.4 1.46 1.89-3.18 3.61-.82-.34-3.68L23 12zm-10 5h-2v-2h2v2zm0-4h-2V7h2v6z"
                />
              </svg>
            </div>
            <span />
          </h2>
        </header>
        <p>
          <span />
        </p>
      </a>
      <script
        type="application/ld+json"
      >
        {"@id":"http://localhost:/news/test-2","@type":"NewsArticle","articleBody":"","author":{"@type":"Person","name":""},"dateModified":"2017-12-31T11:00:00.000Z","datePublished":"2017-12-31T11:00:00.000Z","description":"","headline":"","image":"","mainEntityOfPage":{"@id":"http://localhost:/news/test-2"},"name":"","publisher":{"@type":"LocalBusiness","email":"mgmt@mynameisviii.com","logo":{"@type":"ImageObject","url":"https://s3.amazonaws.com/mynameisviii-static/voodoo-creation-records-logo.png"}},"text":"","thumbnailUrl":"","@context":"http://schema.org/"}
      </script>
    </article>
  </section>
</article>
`;
